extensions/controllers/sandboxwarmpool_controller.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2025 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	extensionsv1alpha1 "sigs.k8s.io/agent-sandbox/extensions/api/v1alpha1"
+)
+
+const (
+	poolLabel = "agents.x-k8s.io/pool"
+)
+
+// SandboxWarmPoolReconciler reconciles a SandboxWarmPool object
+type SandboxWarmPoolReconciler struct {
+	client.Client
+}
+
+//+kubebuilder:rbac:groups=extensions.agents.x-k8s.io,resources=sandboxwarmpools,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=extensions.agents.x-k8s.io,resources=sandboxwarmpools/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile implements the reconciliation loop for SandboxWarmPool
+func (r *SandboxWarmPoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Fetch the SandboxWarmPool instance
+	warmPool := &extensionsv1alpha1.SandboxWarmPool{}
+	if err := r.Get(ctx, req.NamespacedName, warmPool); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.Info("SandboxWarmPool resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get SandboxWarmPool")
+		return ctrl.Result{}, err
+	}
+
+	// Handle deletion
+	if !warmPool.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("SandboxWarmPool is being deleted")
+		return ctrl.Result{}, nil
+	}
+
+	// Save old status for comparison
+	oldStatus := warmPool.Status.DeepCopy()
+
+	// Reconcile the pool (create or delete Pods as needed)
+	if err := r.reconcilePool(ctx, warmPool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Update status if it has changed
+	if err := r.updateStatus(ctx, oldStatus, warmPool); err != nil {
+		log.Error(err, "Failed to update SandboxWarmPool status")
+		return ctrl.Result{}, err
```

Comment:
so i think it is missing retry for this error

Reply:
my bad, ignore. the framwork reconcile if a error returned. I am not using controller-runtime much